### DISCUSSION TITLE
Unreal engine route planners

### DISCUSSION THREAD

**Discussion by nikolatopic131** [carla-simulator/carla#7974] [Q&A]
Hello, I imported a map from Roadrunner into UE following CARLA's instructions (using make import) and even though the opendrive seemed fine in Roadrunner, routes weren't generated for my two on-ramps, so I want to add them manually. Does anybody know how to do that?

I've spawned a route planner on the location where I want my route to begin on the on-ramp, and clicked the "+" on "Traffic routes" to add a route. The spline spawned, but I can't generate extra control points for my spline using alt+LMB (which is how they are generated allegedly) and I checked the keyboard shortcuts in UE and can't find anything related to this. Moreover, for some reason there is a spline generated into the opposite direction of the road that can't be removed (as shown in the picture), while I only want my route to go the way of the road. Also, even when I draw the spline as in the picture, once I de-select it, it disappears. Finally, the automatically generated routes in my map are red, while this manually made spline route looks white (not sure if this fact is of any significance, just curious what the difference is and why is it so).

If someone knows the procedure how to manually create routes in UE, I'd be very grateful if you could share the knowledge.

![Snimka zaslona 2024-07-22 202001](https://github.com/user-attachments/assets/ae849bac-27b6-42d1-95d7-dcc20e66996d)
![Snimka zaslona 2024-07-22 202040](https://github.com/user-attachments/assets/e3acec21-f844-4497-9392-773d9290661b)

**Comment by powderly12** [comment]
@nikolatopic131 Did you find out how to do this? I am facing the same issues

**Reply by nikolatopic131** [reply]
Hey, sorry for the late reply, I saw your question while at work and then forgot to answer once I got home, but by then the notification was no longer there so I forgot completely about it later. I assume you are referring to the issue with the ramps not getting route planners  generated when importing into carla? If so, I figured out that if you built your map in Roadrunner, when importing into carla with make import, it will always disregard the "first" link. So what you need to do is add a tiny link before your ramps in Roadrunner and then the next link after it will have the route planners. However, note that I haven't managed to use on-ramps and off-ramps in my maps because the maps which had them kept giving my agents a routing error (astar path error) that there were some connections missing (even though the map seemed completely fine and had all the route planners). Given these reasons I've opted for a simple bottleneck scenario without any ramps.

I should note though that the routing issues were only when using carla-sumo co-simulation, if you're using standalone carla it should be fine (at least it was for me when I tested).

**Reply by powderly12** [reply]
Hi thank you so much I will give this a try and let you know how it goes

**Reply by powderly12** [reply]
Thank you that worked perfectly!

**Reply by nikolatopic131** [reply]
> Thank you that worked perfectly!

No problem. Glad you resolved your issue.
